worker-message-prioritization-distribution/src/main/java/com/github/workerframework/workermessageprioritization/redistribution/shovel/ShovelStateChecker.java

```diff
@@ -0,0 +1,90 @@
+/*
+ * Copyright 2022-2023 Micro Focus or one of its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.github.workerframework.workermessageprioritization.redistribution.shovel;
+
+import com.github.workerframework.workermessageprioritization.rabbitmq.RabbitManagementApi;
+import com.github.workerframework.workermessageprioritization.rabbitmq.RetrievedShovel;
+import com.github.workerframework.workermessageprioritization.rabbitmq.ShovelState;
+import com.github.workerframework.workermessageprioritization.rabbitmq.ShovelsApi;
+import java.util.Date;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.List;
+
+public final class ShovelStateChecker implements Runnable
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(ShovelStateChecker.class);
+
+    private final RabbitManagementApi<ShovelsApi> shovelsApi;
+    private final String rabbitMQVHost;
+    private final long nonRunningShovelTimeoutMilliseconds;
+
+    public ShovelStateChecker(
+        final RabbitManagementApi<ShovelsApi> shovelsApi,
+        final String rabbitMQVHost,
+        final long nonRunningShovelTimeoutMilliseconds)
+    {
+        this.shovelsApi = shovelsApi;
+        this.rabbitMQVHost = rabbitMQVHost;
+        this.nonRunningShovelTimeoutMilliseconds = nonRunningShovelTimeoutMilliseconds;
+    }
+
+    @Override
+    public void run()
+    {        
+        final List<RetrievedShovel> retrievedShovels = shovelsApi.getApi().getShovels();
+
+        LOGGER.debug("Read the following list of shovels from the RabbitMQ API: {}", retrievedShovels);
+
+        for (final RetrievedShovel retrievedShovel : retrievedShovels) {
+
+            if (retrievedShovel.getState() != ShovelState.RUNNING) {
+
+                final Date shovelCreationTime = retrievedShovel.getTimestamp();
+                final long shovelCreationTimeMilliseconds = shovelCreationTime.getTime();
+
+                final Date timeNow = new Date();
+                final long timeNowMilliseconds = timeNow.getTime();
```

Comment:
I do have a worry about this approach as it would require the clocks to be sync'd. 
I don't see any timezone handling, not sure if it's required.

You might need to have this class record how long it has been observing the shovel in the starting state and use that to determine a fault.

Or maybe I'm being paranoid....

Reply:
Yeah, the RabbitMQ docs say the clock is running in 'local calendar time', so wasn't exactly sure how to handle timezones:

```
Timestamp will return a local calendar time of the form of {{YYYY, MM, DD}, {HH, MM, SS}}.
```
https://www.rabbitmq.com/shovel.html

Regarding the starting state, will need to investigate more, but I'm not sure a shovel can move from a `running` to `starting` state, so as far as I'm aware, the shovel creation time to now **is** how it's been in a starting state.

Reply:
Assigning to @frankmccarry to finish this review.

Frank, I have made the following changes since Andy last reviewed:

**1)**

[Recording the shovel creation time ourselves](https://github.com/WorkerFramework/worker-message-prioritization/pull/4/files#diff-798c13e1bd2ad602660d08133a54d971c4ebc2a648db4cdd7b29620d778a551bR158) using UTC instead of using the time supplied by the RabbitMQ API, which is in a timezone we can't be certain of. This avoids the problems of the clocks being out of synce Andy mentioned above.

**2)**

Added [CAF_WMP_USE_TARGET_QUEUE_CAPACITY_TO_REROUTE](https://github.com/WorkerFramework/worker-message-prioritization/pull/4/files#diff-6353cc97b7c1e4920141251ff248c0e2d8264224918487927b28cc229a80f74dR17) env var to control how we reroute. This allows us to always reroute to a staging queue by setting this to false (default).

Also added a [new RereouteDecider interface and 2 implementation classes](https://github.com/WorkerFramework/worker-message-prioritization/pull/4/files#diff-44b9bc48c8b80177fa247b87dafe71b88d447e735010e6ea7deceb7ecf190c02R70) related to this.